odoo/addons/base/models/ir_profile.py

```diff
@@ -0,0 +1,109 @@
+
+    speedscope = fields.Binary('Speedscope', compute='_compute_speedscope')
+    speedscope_url = fields.Text('Open', compute='_compute_speedscope_url')
+
+    @api.autovacuum
+    def _gc_profile(self):
+        # remove profiles older than 30 days
+        domain = [('create_date', '<', fields.Datetime.now() - datetime.timedelta(days=30))]
+        return self.sudo().search(domain).unlink()
+
+    def _compute_speedscope(self):
+        for execution in self:
+            sp = Speedscope(init_stack_trace=json.loads(execution.init_stack_trace))
+            if execution.sql:
+                sp.add('sql', json.loads(execution.sql))
+            if execution.traces_async:
+                sp.add('frames', json.loads(execution.traces_async))
+            if execution.traces_sync:
+                sp.add('settrace', json.loads(execution.traces_sync))
+
+            result = json.dumps(sp.add_default().make())
+            execution.speedscope = base64.b64encode(result.encode('utf-8'))
+
+    def _compute_speedscope_url(self):
+        for profile in self:
+            profile.speedscope_url = f'/web/speedscope/{profile.id}'
+
+    def _enabled_until(self):
+        """
+        If the profiling is enabled, return until when it is enabled.
+        Otherwise return ``None``.
+        """
+        limit = self.env['ir.config_parameter'].sudo().get_param('base.profiling_enabled_until', '')
+        return limit if str(fields.Datetime.now()) < limit else None
+
+    @api.model
+    def set_profiling(self, profile=None, collectors=None, params=None):
+        """
+        Enable or disable profiling for the current user.
+
+        :param profile: ``True`` to enable profiling, ``False`` to disable it.
+        :param list collectors: optional list of collectors to use (string)
+        :param dict params: optional parameters set on the profiler object
+        """
+        # Note: parameters are coming from a rpc calls or route param (public user),
+        # meaning that corresponding session variables are client-defined.
+        # This allows to activate any profiler, but can be
+        # dangerous handling request.session.profile_collectors/profile_params.
+        if profile:
+            limit = self._enabled_until()
+            _logger.info("User %s started profiling", self.env.user.name)
+            if not limit:
+                request.session.profile_session = None
+                raise UserError(_('Profiling is not enabled on this database'))
+            if not request.session.profile_session:
+                request.session.profile_session = make_session(self.env.user.name)
+                request.session.profile_expiration = limit
+                if request.session.profile_collectors is None:
+                    request.session.profile_collectors = []
+                if request.session.profile_params is None:
+                    request.session.profile_params = {}
+        elif profile is not None:
```

Comment:
this seems to implement "if profile is falsy", but the message that comes across is the opposite? what's the expected value here?

Reply:
None means that we want to keep the current state
any other falsy value means that we want to explicitely disable profiling (False, '', 0)